annealing_helper_functions.py

```diff
@@ -76,3 +77,55 @@ def simulated_annealing(graph, function, initial_X, initial_temp, cool, reanneal
                 T = 1.
             
     return X, history
+
+def prob(a, b):
+    return np.exp(a / b)
+
+# Adapted from Lecture14_Parallel_Tempering_and_Emcee.ipynb
+def parallel_tempering(graph, function, initial_Xs, initial_temps, 
+                       iterr, swap_function, nswaps, nbefore):
+    # Make sure inputs are ok
+    assert(len(initial_temps) == len(initial_Xs)), "Mismatched input dimensions"
+    assert(initial_temps[0] == 1), "First temperature should be one"
+
+    # Initialize stuff
+    nsystems = len(initial_temps)
+    Xs = initial_Xs
+    Ts = initial_temps
+    P = [lambda d: np.exp(d / Ts[c]) for c in range(nsystems)]
+    prev_Es = [function(graph, Xs[i]) for i in range(nsystems)]
+    history = [[] for i in xrange(nsystems)]
+
+    for c in xrange(nsystems):
+        history[c].append(prev_Es[c])
+
+    for step in range(iterr):
+        # Run nbefore steps of simulated annealing
+        X_stars = [swap_function(Xs[i], nswaps) for i in xrange(nsystems)]
+        new_Es = [function(graph, X_stars[i]) for i in xrange(nsystems)]
+        delta_Es = [i - j for i, j in zip(new_Es, prev_Es)]
```

Comment:
Though the formatting is different, this is more-or-less repeated code from simulated annealing. There is probably a way to excise the repeated code into a single helper function called by both simulated annealing and parallel tempering. Also, while the function-in-list structure is supercool, unless I'm mistaken, it will not work for parallelized parallel tempering, so removing it isn't a huge loss.

I'm happy to do the reorg after I'm done with the review, or someone else can do it.